pallets/attesters/src/lib.rs

```diff
@@ -0,0 +1,864 @@
+    }
+
+    #[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+    pub enum AttestationStatus {
+        Pending,
+        Timeout,
+        Approved,
+    }
+
+    #[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+    pub enum BatchStatus {
+        Pending,
+        ReadyForSubmission,
+        Committed,
+    }
+
+    #[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+    pub struct Batch<Attester, Signature, BlockNumber> {
+        pub attestations: Vec<Attestation<Attester, Signature>>,
+        pub target: [u8; 4],
+        pub created: BlockNumber,
+        pub status: BatchStatus,
+    }
+
+    impl<Attester, Signature, BlockNumber: Zero> Default for Batch<Attester, Signature, BlockNumber> {
+        fn default() -> Self {
+            Self {
+                attestations: Vec::new(),
+                target: [0u8; 4],
+                created: Zero::zero(),
+                status: BatchStatus::Pending,
+            }
+        }
+    }
+
+    impl<Attester, Signature, BlockNumber> Batch<Attester, Signature, BlockNumber> {
+        pub fn new(target: [u8; 4], now: BlockNumber) -> Self {
+            Self {
+                attestations: Vec::new(),
+                target,
+                created: now,
+                status: BatchStatus::Pending,
+            }
+        }
+    }
+
+    #[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+    pub struct Attestation<Attester, Signature> {
+        pub for_: AttestationFor,
+        pub status: AttestationStatus,
+        pub signature: Vec<(Attester, Signature)>,
+    }
+
+    impl<Attester, Signature> Default for Attestation<Attester, Signature> {
+        fn default() -> Self {
+            Self {
+                for_: AttestationFor::SFX,
+                status: AttestationStatus::Pending,
+                signature: Vec::new(),
+            }
+        }
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        type ActiveSetSize: Get<u32>;
+        type CommitteeSize: Get<u32>;
+        type BatchingWindow: Get<Self::BlockNumber>;
+        type ShufflingFrequency: Get<Self::BlockNumber>;
+        type MaxBatchSize: Get<u32>;
+        type RewardMultiplier: Get<BalanceOf<Self>>;
+        type CommitmentRewardSource: Get<Self::AccountId>;
+        type Currency: ReservableCurrency<Self::AccountId>;
+        type RandomnessSource: Randomness<Self::Hash, Self::BlockNumber>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn attesters)]
+    pub type Attesters<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, AttesterInfo>;
+
+    #[pallet::storage]
+    pub type CurrentCommittee<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
```

Comment:
We should consider doing this index based. This makes the signature verification on the target a lot cheaper, as we can get a pub key based on the index, instead of looping. This was done in Ethereum and we just need to sort the committee keys based on some condition when initializing and updating the committee